File: Dev/C++/Mascaret/src/VEHA/Behavior/Activity/ActivityExecution.cpp
#include "VEHA/Behavior/Activity/ActivityExecution.h"

namespace VEHA
{

ActivityExecution::ActivityExecution(const shared_ptr<Activity> act) :
		_activity(act)
{
	_toExecuteNodes = act->getInitialNode()->getOutgoingActionNode();
}

ActivityExecution::~ActivityExecution()
{
}

vector<shared_ptr<ActionNode> > ActivityExecution::getNextActions(const string& roleName)
{
	vector<shared_ptr<ActionNode> > toExec;
	updateNodes();
	for(size_t i=0;i<_toExecuteNodes.size();i++)
	{
		if(_toExecuteNodes[i]->getPartition_at(0)->getName() == roleName)
		{
			toExec.push_back(_toExecuteNodes[i]);
		}
	}
	return toExec;
}
string clean(const string& toClean)
{
	string returned="";
	for(size_t i=0;i<toClean.size();i++)
	{
		if(toClean[i] != '/' && toClean[i] != '(' && toClean[i] != ')' && toClean[i] != '^')
			returned+=toClean[i];
	}
	return returned;
}

void ActivityExecution::printState()
{
	cerr << "Noeud a executer:"<<endl;
	vector<shared_ptr<ActionNode> >::iterator it=_toExecuteNodes.begin();
	while(it != _toExecuteNodes.end())
	{
		cerr << "--> "<<(*it)->getName() << " par "<< (*it)->getPartition_at(0)->getName() << endl;
		it++;
	}
	cerr << "Noeud en cours d'execution:"<<endl;
	it=_runningNodes.begin();
	while(it != _runningNodes.end())
	{
		cerr << "--> " << (*it)->getName() << " par "<< (*it)->getPartition_at(0)->getName() << endl;
		it++;
	}
	cerr << "Evenement running recu:"<<endl;
	vector<pair<string,string> >::iterator it2=_actionsRunning.begin();
	while(it2 != _actionsRunning.end())
	{
		cerr << "--> "<<it2->first << " running " << it2->second << endl;
		it2++;
	}
	cerr << "Evenement done recu:"<<endl;
	it2=_actionsDone.begin();
	while(it2 != _actionsDone.end())
	{
		cerr << "--> "<<it2->first << " have done "<<it2->second << endl;
		it2++;
	}

}
template <typename T> inline
void
StlVectorFastErase(vector<T> & vect,
		unsigned int index)
{
	vect[index]=vect.back();
	vect.pop_back();
}
void ActivityExecution::updateNodes()
{
	for(size_t i=0;i<_actionsRunning.size();)
	{
		bool increase=true;
		for(size_t j=0;j<_toExecuteNodes.size();)
		{
			if(_actionsRunning[i].first == _toExecuteNodes[j]->getPartition_at(0)->getName() &&
					_actionsRunning[i].second == clean(_toExecuteNodes[j]->getName()))
			{
				//cerr << _toExecuteNodes[j]->getName() << " en cours d'execution"<<endl;
				_runningNodes.push_back(_toExecuteNodes[j]);
				StlVectorFastErase(_toExecuteNodes,j);
				StlVectorFastErase(_actionsRunning,i);
				increase=false;
				break;
			}
			else
			{
				j++;
			}
		}
		if(increase)
			i++;
	}

	for(size_t i=0;i<_actionsDone.size();)
	{
		bool increase=true;
		for(size_t j=0;j<_runningNodes.size();)
		{
			if(_actionsDone[i].first == _runningNodes[j]->getPartition_at(0)->getName() &&
					_actionsDone[i].second == clean(_runningNodes[j]->getName()))
			{
				//cerr << _runningNodes[j]->getName() << " terminé"<<endl;
				//cerr << "find next of "<< _runningNodes[j]->getName() <<endl;
				vector< shared_ptr < ActionNode> > next = _runningNodes[j]->getOutgoingActionNode();
				for (size_t iN = 0; iN < next.size() ; iN++)
				{
					//cerr << _runningNodes[j]->getName()<<" nextNode: "<<next[iN]->getName()<<endl;
					_toExecuteNodes.push_back(next[iN]);
				}
				StlVectorFastErase(_runningNodes,j);
				StlVectorFastErase(_actionsDone,i);
				break;
			}
			else
			{
				j++;
			}
		}
		if(increase)
			i++;
	}
}

}

File: Dev/CS/Mascaret/Mascaret/VEHA/Behavior/Activity/ActivityExecution.cs
using System;
using System.Collections.Generic;

namespace Mascaret
{
    public class ActivityExecution
    {

        private List<ActivityNode> runningNodes = new List<ActivityNode>();
        public List<ActivityNode> RunningNodes
        {
            get { return runningNodes; }
            set { runningNodes = value; }
        }


        private List<ActionNode> toExecetuteNodes = new List<ActionNode>();
        public List<ActionNode> ToExecetuteNodes
        {
            get { return toExecetuteNodes; }
            set { toExecetuteNodes = value; }
        }


        private List<KeyValuePair<string, string>> actionsDone = new List<KeyValuePair<string, string>>();
        public List<KeyValuePair<string, string>> ActionsDone
        {
            get { return actionsDone; }
            set { actionsDone = value; }
        }


        private List<KeyValuePair<string, string>> actionsRunning = new List<KeyValuePair<string, string>>();
        public List<KeyValuePair<string, string>> ActionsRunning
        {
            get { return actionsRunning; }
            set { actionsRunning = value; }
        }


        private Activity activity;
        public Activity Activity
        {
            get { return activity; }
            set { activity = value; }
        }

        public ActivityExecution(Activity activity)
        {
            this.activity = activity;
            toExecetuteNodes = activity.Initial.getOutgoingActionNode();
        }

        public List<ActionNode> getNextActions(string roleName)
        {
            List<ActionNode> toExec = new List<ActionNode>();
            updateNodes();
            foreach (ActionNode currentNode in toExecetuteNodes)
            {
                if (currentNode.Partitions[0].name == roleName)
                    toExec.Add(currentNode);
            }

            return toExec;
        }

        //Ce n'est peut être pas très joli mais au moins ça fait ce qu'il faut :D
        protected string clean(string toClean)
        {
            return toClean.Replace("/", "").Replace("(", "").Replace(")", "").Replace("^", "");
        }

        public bool isFinished()
        {
            return (toExecetuteNodes.Count == 0 && runningNodes.Count == 0);
        }

        public void informActionRunning(string roleName, string actionName)
        {
            actionsRunning.Add(new KeyValuePair<string, string>(roleName, actionName));
        }

        public void informActionDone(string roleName, string actionName)
        {
            actionsDone.Add(new KeyValuePair<string, string>(roleName, actionName));
        }

        public void printState()
        {
            System.Console.WriteLine("Noeud a executer : ");
            foreach (ActionNode currentNode in toExecetuteNodes)
            {
                System.Console.WriteLine("-->" + currentNode.name + " par " + currentNode.Partitions[0].name);
            }
            System.Console.WriteLine("Noeud en cours d'execution : ");
            foreach (ActionNode currentNode in runningNodes)
            {
                System.Console.WriteLine("-->" + currentNode.name + " par " + currentNode.Partitions[0].name);
            }
            System.Console.WriteLine("Evenement running recu : ");
            foreach (KeyValuePair<string, string> currentNode in actionsRunning)
            {
                System.Console.WriteLine("-->" + currentNode.Key + " running " + currentNode.Value);
            }
            System.Console.WriteLine("Evenement done recu : ");
            foreach (KeyValuePair<string, string> currentNode in actionsDone)
            {
                System.Console.WriteLine("-->" + currentNode.Key + " have done " + currentNode.Value);
            }
        }

        public void updateNodes()
        {
            foreach (KeyValuePair<string, string> currentPair in actionsRunning)
            {
                int i = 0, j = 0;
                bool increase = true;
                foreach (ActionNode currentNode in toExecetuteNodes)
                {
                    if (currentPair.Key == currentNode.Partitions[0].name && currentPair.Value == clean(currentNode.name))
                    {
                        runningNodes.Add(currentNode);
                        toExecetuteNodes.RemoveRange(j, 1);
                        actionsRunning.RemoveRange(i, 1);
                        increase = false;
                        break;
                    }
                    else
                        j++;
                    j++;
                }
                if (increase)
                    i++;
                i++;
            }

            foreach (KeyValuePair<string, string> currentPair in actionsDone)
            {
                int i = 0, j = 0;
                bool increase = true;
                foreach (ActionNode currentNode in runningNodes)
                {
                    if (currentPair.Key == currentNode.Partitions[0].name && currentPair.Value == clean(currentNode.name))
                    {
                        runningNodes.RemoveRange(j, 1);
                        actionsDone.RemoveRange(i, 1);
                        break;
                    }
                    else
                        j++;
                    j++;
                }
                if (increase)
                    i++;
                i++;
            }

        }


    }
}
